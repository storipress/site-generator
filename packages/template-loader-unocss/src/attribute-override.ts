import { extractBlockId } from './block-id'
import type { Breakpoint } from './convert-attribute'
import { articleConverter, frontPageConverter, isPathShouldHoist, pathToSelector } from './path-converter'
import type { AttributeContext } from './transform'
import type { StyleTree } from './types'

interface HoistConfig {
  hoistAs: string
  keys: Set<string>
}
const SHOULD_HOIST = new Map<string, HoistConfig>([
  [
    'SpacingProvider',
    {
      hoistAs: '.spacing',
      keys: new Set(['width', 'maxWidth', 'minWidth', 'max', 'min']),
    },
  ],
])

interface TagAliasConfig {
  type?: 'front' | 'article'
  aliases: string[]
}

const TAG_ALIASES = new Map<string, TagAliasConfig>([
  [
    'Paragraph',
    {
      aliases: [':deep(.base-text)'],
    },
  ],
  [
    'Authors',
    {
      type: 'article',
      aliases: [':deep(.author-name)'],
    },
  ],
])

interface CreateAttributeOverrideInput {
  root: string
  map: Record<string, Record<string, Record<string, Breakpoint>>> | StyleTree
  get?: any
}

const converters = {
  front: frontPageConverter,
  article: articleConverter,
}

const ALLOW_NO_KIND = new Set(['Block', 'HeroBlock', 'FooterBlock', 'ColorArea', ...SHOULD_HOIST.keys()])

/**
 * create plugin to do override styles to merge user defined styles
 * Make this a plugin to make it easier to test
 * @param root
 * @param map
 * @returns
 */
export function createAttributeOverride({
  root,
  map,
  get = getStyles,
}: CreateAttributeOverrideInput): (ctx: AttributeContext) => void {
  return (ctx) => {
    const res = extractBlockId(ctx.id, root)

    if (res?.type === 'front' && !ctx.kind && !ALLOW_NO_KIND.has(ctx.tag)) {
      return
    }

    if (res) {
      const { type, blockId } = res
      const convert = converters[type]
      const path = convert({
        tag: ctx.tag,
        id: blockId,
        kind: ctx.kind,
      })
      if (path) {
        const override = get(map, path)?.[ctx.key]
        if (override) {
          ctx.value = {
            ...ctx.value,
            ...override,
          }
        }

        if (isPathShouldHoist(path)) {
          const scopedHoistAs = ctx.scopedHoistAs ? [...ctx.scopedHoistAs] : []
          ctx.scopedHoistAs = [...scopedHoistAs, pathToSelector(path)]
        }

        const tagAliasConfig = TAG_ALIASES.get(ctx.tag)
        if (tagAliasConfig && (!tagAliasConfig.type || tagAliasConfig.type === type)) {
          const scopedHoistAs = ctx.scopedHoistAs ? [...ctx.scopedHoistAs] : []
          ctx.scopedHoistAs = [...scopedHoistAs, ...tagAliasConfig.aliases]
        }
      }
    }

    const hoistConfig = SHOULD_HOIST.get(ctx.tag)
    if (hoistConfig && hoistConfig.keys.has(ctx.key)) {
      ctx.hoistAs = hoistConfig.hoistAs
    }
  }
}

function getStyles(obj: StyleTree, path: string[]): any {
  const lastKey = path[path.length - 1]
  let res: any = obj
  for (const key of path) {
    if (!res[key]) return
    if (res[lastKey]) return res[lastKey].styles

    res = res[key] ? res[key].children : {}
  }
}
