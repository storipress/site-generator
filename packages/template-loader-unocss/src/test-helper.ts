import uno from '@unocss/preset-uno'
import { h } from '@unocss/preset-mini/utils'
import unocss from 'unocss/vite'
import { build } from 'vite'
import type { OutputAsset, RollupOutput } from 'rollup'
import transformerDirectives from '@unocss/transformer-directives'
import vue from '@vitejs/plugin-vue'
import type { NestedHooks } from 'hookable'
import transformer from './transform'
import type { TransformerHooks } from './transform'

interface CreateTransformBuildInput {
  code: string
  pathPrefix?: string
  hooks?: NestedHooks<TransformerHooks>
}

export async function createTransformBuild({ code, pathPrefix, hooks }: CreateTransformBuildInput) {
  let vueCode = ''
  const res = await build({
    logLevel: 'error',
    configFile: false,
    plugins: [
      {
        name: 'virtual',
        resolveId(id) {
          if (id.endsWith('entry')) {
            return id
          }
          if (id.endsWith('entry.vue')) {
            return id
          }
        },
        load(id) {
          if (id.endsWith('entry')) {
            return `
              export {default} from '${pathPrefix ?? ''}/entry.vue'
            `
          }
          if (id.endsWith('entry.vue')) {
            return code
          }
        },
      },
      unocss({
        mode: 'vue-scoped',
        presets: [
          uno({
            preflight: false,
          }),
        ],
        rules: [[/^ff-(.+)$/, ([, family]) => ({ 'font-family': `"${h.bracket(family)}"` })]],
        transformers: [
          transformer({
            hooks,
          }),
          transformerDirectives(),
        ],
      }),
      {
        name: 'inspect',
        transform(code, id) {
          if (id.endsWith('entry.vue')) {
            vueCode = code
          }
          return code
        },
      },
      vue({}),
    ],
    build: {
      lib: {
        entry: 'entry',
        formats: ['es'],
      },
      rollupOptions: {
        output: undefined,
        external: ['vue'],
      },
    },
  })

  const out = res as RollupOutput[]

  return {
    code: out[0].output[0].code,
    vue: vueCode,
    css: (out[0].output[1] as OutputAsset)?.source || '',
  }
}
