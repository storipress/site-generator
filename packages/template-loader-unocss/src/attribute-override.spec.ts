import { expect, it } from 'vitest'
import { html } from 'proper-tags'
import { createTransformBuild } from './test-helper'
import { createAttributeOverride } from './attribute-override'

function get(obj: Record<string, any>, path: string[]): any {
  let res = obj
  for (const key of path) {
    res = res[key]
    if (!res) return
  }
  return res
}

it('createAttributeOverride: basic', async () => {
  const { code } = await createTransformBuild({
    code: html`
      <template>
        <TextInput kind="foo" fontSize="12" color="ff0000" />
      </template>
    `,
    pathPrefix: '/blocks/front/block-id',
    hooks: {
      'transformer:attribute': createAttributeOverride({
        get,
        root: '/blocks',
        map: {
          'b-block-id': {
            foo: {
              fontSize: {
                xs: '16',
              },
            },
          },
        },
      }),
    },
  })
  expect(code).toMatch('text-[16px]')
})

it('createAttributeOverride: hoist', async () => {
  const { vue, css } = await createTransformBuild({
    code: html`
      <template>
        <SpacingProvider width="80%" />
      </template>
    `,
    pathPrefix: '/blocks/front/block-id',
    hooks: {
      transformer: {
        attribute: createAttributeOverride({ get, root: '/blocks', map: {} }),
      },
    },
  })
  expect(vue).toMatchSnapshot()
  expect(css).toMatchSnapshot()
})

it('createAttributeOverride: hoist multiple', async () => {
  const { vue, css } = await createTransformBuild({
    code: html`
      <template>
        <SpacingProvider width="80%" max="1200" />
      </template>
    `,
    pathPrefix: '/blocks/front/block-id',
    hooks: {
      'transformer:attribute': createAttributeOverride({ get, root: '/blocks', map: {} }),
    },
  })
  expect(vue).toMatchSnapshot()
  expect(css).toMatchSnapshot()
})

it('createAttributeOverride: article scoped hoist', async () => {
  const { vue, css } = await createTransformBuild({
    code: html`
      <template>
        <ArticleBlock>
          <ArticleContent>
            <Paragraph color="ff0000" />
          </ArticleContent>
        </ArticleBlock>
      </template>
    `,
    pathPrefix: '/blocks/article/block-id',
    hooks: {
      'transformer:attribute': createAttributeOverride({ get, root: '/blocks', map: {} }),
    },
  })
  expect(vue).toMatchSnapshot()
  expect(css).toMatchSnapshot()
})

it('createAttributeOverride: article Authors', async () => {
  const { vue, css } = await createTransformBuild({
    code: html`
      <template>
        <ArticleBlock>
          <Authors bold />
        </ArticleBlock>
      </template>
    `,
    pathPrefix: '/blocks/article/block-id',
    hooks: {
      'transformer:attribute': createAttributeOverride({ get, root: '/blocks', map: {} }),
    },
  })
  expect(vue).toMatchSnapshot()
  expect(css).toMatchSnapshot()
})

it('createAttributeOverride: article Paragraph', async () => {
  const { vue, css } = await createTransformBuild({
    code: html`
      <template>
        <ArticleBlock>
          <Paragraph bold />
        </ArticleBlock>
      </template>
    `,
    pathPrefix: '/blocks/article/block-id',
    hooks: {
      'transformer:attribute': createAttributeOverride({ get, root: '/blocks', map: {} }),
    },
  })
  expect(vue).toMatchSnapshot()
  expect(css).toMatchSnapshot()
})
