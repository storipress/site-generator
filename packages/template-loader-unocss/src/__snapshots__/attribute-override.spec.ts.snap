// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createAttributeOverride: article Authors 1`] = `
"<template>
  <ArticleBlock>
    <Authors      />
  </ArticleBlock>
</template>
<style lang=\\"postcss\\" scoped>
  :deep(.author-name) {
            @apply font-bold;
          }
</style>

<style scoped>/* layer: default */
.font-bold{font-weight:700;}</style>"
`;

exports[`createAttributeOverride: article Authors 2`] = `
"[data-v-scopeid] .author-name,.font-bold[data-v-scopeid]{font-weight:700}
"
`;

exports[`createAttributeOverride: article Paragraph 1`] = `
"<template>
  <ArticleBlock>
    <Paragraph      />
  </ArticleBlock>
</template>
<style lang=\\"postcss\\" scoped>
  .article .article-content :deep(.main-content p) {
            @apply font-bold;
          }


          :deep(.base-text) {
            @apply font-bold;
          }
</style>

<style scoped>/* layer: default */
.font-bold{font-weight:700;}</style>"
`;

exports[`createAttributeOverride: article Paragraph 2`] = `
".article .article-content[data-v-scopeid] .main-content p,[data-v-scopeid] .base-text,.font-bold[data-v-scopeid]{font-weight:700}
"
`;

exports[`createAttributeOverride: article scoped hoist 1`] = `
"<template>
  <ArticleBlock>
    <ArticleContent>
      <Paragraph                />
    </ArticleContent>
  </ArticleBlock>
</template>
<style lang=\\"postcss\\" scoped>
  .article .article-content :deep(.main-content p) {
            @apply text-[#ff0000];
          }


          :deep(.base-text) {
            @apply text-[#ff0000];
          }
</style>

<style scoped>/* layer: default */
.text-\\\\[\\\\#ff0000\\\\]{--un-text-opacity:1;color:rgb(255 0 0 / var(--un-text-opacity));}</style>"
`;

exports[`createAttributeOverride: article scoped hoist 2`] = `
".article .article-content[data-v-scopeid] .main-content p,[data-v-scopeid] .base-text,.text-\\\\[\\\\#ff0000\\\\][data-v-scopeid]{--un-text-opacity:1;color:rgb(255 0 0 / var(--un-text-opacity))}
"
`;

exports[`createAttributeOverride: hoist 1`] = `
"<template>
  <SpacingProvider             />
</template>
<style lang=\\"postcss\\" >
  .spacing {
            @apply w-[80%];
          }
</style>

<style scoped>/* layer: default */
.w-\\\\[80\\\\%\\\\]{width:80%;}</style>"
`;

exports[`createAttributeOverride: hoist 2`] = `
".spacing,.w-\\\\[80\\\\%\\\\][data-v-scopeid]{width:80%}
"
`;

exports[`createAttributeOverride: hoist multiple 1`] = `
"<template>
  <SpacingProvider                        />
</template>
<style lang=\\"postcss\\" >
  .spacing {
            @apply w-[80%] max-w-[1200px];
          }
</style>

<style scoped>/* layer: default */
.max-w-\\\\[1200px\\\\]{max-width:1200px;}
.w-\\\\[80\\\\%\\\\]{width:80%;}</style>"
`;

exports[`createAttributeOverride: hoist multiple 2`] = `
".spacing{width:80%;max-width:1200px}.max-w-\\\\[1200px\\\\][data-v-scopeid]{max-width:1200px}.w-\\\\[80\\\\%\\\\][data-v-scopeid]{width:80%}
"
`;
